sdk/java/pulumi/src/main/java/com/pulumi/core/Output.java

```
@@ -259,6 +243,20 @@ private static Output<Object> ensureOutput(@Nullable Object o) {
         return Output.ofNullable(o);
     }
 
+    private static <T> OutputData<T> joinOrThrow(CompletableFuture<OutputData<T>> futureData) {
+        var d = futureData.join();
+        return requireNonNull(d, "unexpected null OutputData from a CompletableFuture");
```

throw on regression or bug instead of swallowing the null